base/linalg/matmul.jl

```diff
@@ -424,7 +426,7 @@ end
 function generic_matmatmul{T,S}(tA, tB, A::AbstractVecOrMat{T}, B::AbstractMatrix{S})
     mA, nA = lapack_size(tA, A)
     mB, nB = lapack_size(tB, B)
-    C = similar(B, promote_op(*, arithtype(T), arithtype(S)), mA, nB)
+    C = similar(B, promote_op(matprod, arithtype(T), arithtype(S)), mA, nB)
```

Comment:
I guess they just were unintentionally forgot. But, now that you mention it, `matprod` also solves (more generally) the problem they where supposed to tackle, so in fact we can remove them and their definitions.